api/src/main/java/com/epam/pipeline/manager/datastorage/providers/aws/s3/S3Helper.java

```diff
@@ -300,6 +303,44 @@ public void restoreFileVersion(String bucket, String path, String version) {
         moveS3Object(client, bucket, new MoveObjectRequest(path, version, path));
     }
 
+    public void restoreFolderVersion(String bucket, String path, RestoreFolderVO restoreFolderVO) {
+        AmazonS3 client = getDefaultS3Client();
+        try (S3ObjectDeleter deleter = new S3ObjectDeleter(client, bucket)) {
+            getDeletedItems(bucket, path, true, null, null, restoreFolderVO)
+                    .forEach(item -> deleter.deleteKey(item.getPath(), ((DataStorageFile) item).getVersion()));
+        } catch (SdkClientException e) {
+            throw new DataStorageException(e.getMessage(), e.getCause());
+        }
+    }
+
+    public List<AbstractDataStorageItem> getDeletedItems(String bucket, String path, Boolean showVersion,
+                                                         Integer pageSize, String marker, RestoreFolderVO restoreFolderVO) {
+        String requestPath = Optional.ofNullable(path).orElse("");
+        AmazonS3 client = getDefaultS3Client();
+        if (!StringUtils.isNullOrEmpty(requestPath)) {
+            DataStorageItemType type = checkItemType(client, bucket, requestPath, showVersion);
+            if (type == DataStorageItemType.Folder && !requestPath.endsWith(ProviderUtils.DELIMITER)) {
+                requestPath += ProviderUtils.DELIMITER;
+            }
+        }
+        return listVersions(client, bucket, requestPath, pageSize, marker).getResults().stream()
+                .peek(item -> {
+                    if (item.getType() == DataStorageItemType.Folder && restoreFolderVO.isRecursively()) {
+                        restoreFolderVersion(bucket, item.getPath(), restoreFolderVO);
+                    }
+                })
+                .filter(i -> i.getType() == DataStorageItemType.File)
```

Comment:
Could we place all filters into one condition?
Also, to my mind, the names of variables should be more human readable.

Reply:
fixed